File: android/app/src/main/java/com/example/app/domain/navcaddy/offline/OfflineCapability.kt
package caddypro.domain.navcaddy.offline

import caddypro.domain.navcaddy.models.IntentType

/**
 * Defines which intents are available in offline mode.
 *
 * Offline-capable intents can be processed locally without LLM API calls.
 * These intents typically involve local data operations (score entry, stats lookup)
 * or navigation to local screens (settings, equipment).
 *
 * Spec reference: navcaddy-engine.md C6 (Offline behavior), NG4 (limited local intents only)
 * Plan reference: navcaddy-engine-plan.md Task 24
 */
object OfflineCapability {

    /**
     * Set of intents that can be processed offline.
     *
     * These intents do not require LLM classification or external API calls.
     * They can be handled with local pattern matching and data operations.
     *
     * Per spec C6: "If offline: disable LLM calls and provide a limited local-intent menu"
     */
    val OFFLINE_AVAILABLE_INTENTS = setOf(
        // Score operations - local database only
        IntentType.SCORE_ENTRY,

        // Stats - read from local database
        IntentType.STATS_LOOKUP,

        // Equipment - local data
        IntentType.EQUIPMENT_INFO,

        // Round management - local state
        IntentType.ROUND_START,
        IntentType.ROUND_END,

        // Settings - local preferences
        IntentType.SETTINGS_CHANGE,

        // Help - static content
        IntentType.HELP_REQUEST,

        // Club adjustment - local data update
        IntentType.CLUB_ADJUSTMENT,

        // Pattern query - local database
        IntentType.PATTERN_QUERY
    )

    /**
     * Intents that require online connectivity.
     *
     * These intents need LLM classification, external APIs, or real-time data.
     */
    val ONLINE_REQUIRED_INTENTS = setOf(
        // Shot recommendation - requires LLM and context
        IntentType.SHOT_RECOMMENDATION,

        // Recovery - may require external data
        IntentType.RECOVERY_CHECK,

        // Drill request - requires LLM for personalization
        IntentType.DRILL_REQUEST,

        // Weather - external API required
        IntentType.WEATHER_CHECK,

        // Course info - may require external data
        IntentType.COURSE_INFO,

        // Feedback - requires network to submit
        IntentType.FEEDBACK
    )

    /**
     * Check if an intent can be processed offline.
     *
     * @param intentType The intent to check
     * @return true if intent can be processed without network, false otherwise
     */
    fun isOfflineAvailable(intentType: IntentType): Boolean {
        return intentType in OFFLINE_AVAILABLE_INTENTS
    }

    /**
     * Check if an intent requires online connectivity.
     *
     * @param intentType The intent to check
     * @return true if intent requires network, false otherwise
     */
    fun requiresOnline(intentType: IntentType): Boolean {
        return intentType in ONLINE_REQUIRED_INTENTS
    }

    /**
     * Get user-friendly message explaining offline limitations.
     *
     * @param intentType The intent that was attempted offline
     * @return Message explaining why intent is unavailable offline
     */
    fun getOfflineLimitationMessage(intentType: IntentType): String {
        return when (intentType) {
            IntentType.SHOT_RECOMMENDATION ->
                "Shot recommendations need an internet connection. Try checking your stats or equipment instead."

            IntentType.RECOVERY_CHECK ->
                "Recovery insights need an internet connection. Check back when you're online."

            IntentType.DRILL_REQUEST ->
                "Personalized drills need an internet connection. Check your patterns in the meantime."

            IntentType.WEATHER_CHECK ->
                "Weather data needs an internet connection. I can't check conditions offline."

            IntentType.COURSE_INFO ->
                "Course information needs an internet connection. Try looking at your saved rounds instead."

            IntentType.FEEDBACK ->
                "Feedback submission needs an internet connection. Your feedback will be saved and sent when you're back online."

            else ->
                "This feature needs an internet connection. You can still enter scores, check stats, or view your equipment."
        }
    }

    /**
     * Get list of offline-available intents for user presentation.
     *
     * Returns a prioritized list of the most useful offline intents.
     *
     * @return List of intents available offline, in priority order
     */
    fun getOfflineIntentsPrioritized(): List<IntentType> {
        return listOf(
            IntentType.SCORE_ENTRY,
            IntentType.STATS_LOOKUP,
            IntentType.EQUIPMENT_INFO,
            IntentType.PATTERN_QUERY,
            IntentType.CLUB_ADJUSTMENT,
            IntentType.ROUND_START,
            IntentType.ROUND_END,
            IntentType.SETTINGS_CHANGE,
            IntentType.HELP_REQUEST
        )
    }

    /**
     * Get general offline mode message for users.
     *
     * @return User-friendly message explaining offline mode
     */
    fun getOfflineModeMessage(): String {
        return "You're offline. I can help with scores, stats, equipment, and settings. " +
                "Full features will be back when you reconnect."
    }

    /**
     * Get online reconnection message for users.
     *
     * @return User-friendly message when connection is restored
     */
    fun getOnlineModeMessage(): String {
        return "You're back online! All features are now available."
    }
}

File: ios/App/Sources/Core/Domain/NavCaddy/Offline/OfflineCapability.swift
import Foundation

/**
 * Defines which intents can be handled offline.
 *
 * Offline-capable intents are those that:
 * - Don't require LLM processing
 * - Use only local data (SwiftData, UserDefaults)
 * - Have deterministic routing
 *
 * Spec reference: navcaddy-engine.md C6 (Offline behavior)
 * Plan reference: navcaddy-engine-plan.md Task 24
 */
struct OfflineCapability {

    // MARK: - Offline Intent List

    /// Intents that can be handled offline.
    ///
    /// These intents are supported without network connectivity:
    /// - SCORE_ENTRY: Log scores locally
    /// - STATS_LOOKUP: View local statistics
    /// - EQUIPMENT_INFO: View club distances
    /// - ROUND_START: Start round tracking
    /// - ROUND_END: End round and save locally
    /// - SETTINGS_CHANGE: Update app settings
    /// - HELP_REQUEST: View static help content
    /// - CLUB_ADJUSTMENT: Adjust club distances locally
    /// - PATTERN_QUERY: View local pattern data
    static let offlineCapableIntents: Set<IntentType> = [
        .scoreEntry,
        .statsLookup,
        .equipmentInfo,
        .roundStart,
        .roundEnd,
        .settingsChange,
        .helpRequest,
        .clubAdjustment,
        .patternQuery
    ]

    // MARK: - Capability Checks

    /// Check if an intent can be handled offline.
    static func isOfflineCapable(_ intentType: IntentType) -> Bool {
        offlineCapableIntents.contains(intentType)
    }

    /// Get the routing target for an offline-capable intent.
    ///
    /// Returns nil if the intent is not offline-capable or requires
    /// additional context to determine the target.
    static func routingTarget(for intentType: IntentType) -> RoutingTarget? {
        guard isOfflineCapable(intentType) else {
            return nil
        }

        switch intentType {
        case .scoreEntry:
            return RoutingTarget(
                module: .caddy,
                screen: "score_entry",
                parameters: [:]
            )

        case .statsLookup:
            return RoutingTarget(
                module: .caddy,
                screen: "stats",
                parameters: [:]
            )

        case .equipmentInfo:
            return RoutingTarget(
                module: .caddy,
                screen: "equipment",
                parameters: [:]
            )

        case .roundStart:
            return RoutingTarget(
                module: .caddy,
                screen: "round_start",
                parameters: [:]
            )

        case .roundEnd:
            return RoutingTarget(
                module: .caddy,
                screen: "round_summary",
                parameters: [:]
            )

        case .settingsChange:
            return RoutingTarget(
                module: .settings,
                screen: "main",
                parameters: [:]
            )

        case .helpRequest:
            return RoutingTarget(
                module: .settings,
                screen: "help",
                parameters: [:]
            )

        case .clubAdjustment:
            return RoutingTarget(
                module: .caddy,
                screen: "club_adjustment",
                parameters: [:]
            )

        case .patternQuery:
            return RoutingTarget(
                module: .caddy,
                screen: "patterns",
                parameters: [:]
            )

        default:
            return nil
        }
    }

    // MARK: - User Messages

    /// Get user-facing message for offline capability.
    static func offlineMessage(for intentType: IntentType) -> String {
        switch intentType {
        case .scoreEntry:
            return "You can enter your score offline. It will sync when you're back online."

        case .statsLookup:
            return "Here are your stats. Some data might not be current while offline."

        case .equipmentInfo:
            return "Here's your bag setup. Changes will sync when you're back online."

        case .roundStart:
            return "Starting your round. Tracking will continue offline."

        case .roundEnd:
            return "Round complete! Your data will sync when you're back online."

        case .settingsChange:
            return "Settings updated. Some features may be limited while offline."

        case .helpRequest:
            return "Here's some help for you. Full documentation is available when you're online."

        case .clubAdjustment:
            return "Adjusting your club distances. Changes will sync when you're back online."

        case .patternQuery:
            return "Here are your patterns from local data. Full analysis available when you're online."

        default:
            return "This action is available offline."
        }
    }

    /// Get message explaining why an intent is not available offline.
    static func unavailableMessage(for intentType: IntentType) -> String {
        switch intentType {
        case .shotRecommendation:
            return "Shot recommendations need a connection to provide personalized advice."

        case .recoveryCheck:
            return "Recovery checks need a connection to analyze your data."

        case .drillRequest:
            return "Drill recommendations need a connection."

        case .weatherCheck:
            return "Weather information requires a connection."

        case .courseInfo:
            return "Course information requires a connection."

        case .feedback:
            return "Feedback will be sent when you're back online."

        default:
            return "This feature requires a network connection."
        }
    }
}

// MARK: - Offline Mode Configuration

/// Configuration for offline mode behavior.
struct OfflineModeConfiguration {
    /// Whether to show offline indicator in UI.
    let showOfflineIndicator: Bool

    /// Whether to queue online-only requests for when connection returns.
    let queueOnlineRequests: Bool

    /// Whether to show limited functionality warnings.
    let showLimitedFunctionalityWarnings: Bool

    /// Default configuration.
    static let `default` = OfflineModeConfiguration(
        showOfflineIndicator: true,
        queueOnlineRequests: false, // Don't auto-queue for privacy
        showLimitedFunctionalityWarnings: true
    )
}

// MARK: - Offline Suggestions

extension OfflineCapability {
    /// Get suggested offline-capable intents.
    static func offlineSuggestions() -> [IntentSuggestion] {
        [
            IntentSuggestion(
                intentType: .scoreEntry,
                label: "Enter score",
                description: "Log your score for this hole"
            ),
            IntentSuggestion(
                intentType: .equipmentInfo,
                label: "View my bag",
                description: "See your club setup"
            ),
            IntentSuggestion(
                intentType: .statsLookup,
                label: "View stats",
                description: "Check your performance statistics"
            )
        ]
    }

    /// Get context-aware offline suggestions based on round state.
    static func contextAwareOfflineSuggestions(isRoundActive: Bool) -> [IntentSuggestion] {
        if isRoundActive {
            return [
                IntentSuggestion(
                    intentType: .scoreEntry,
                    label: "Enter score",
                    description: "Log your score for this hole"
                ),
                IntentSuggestion(
                    intentType: .roundEnd,
                    label: "End round",
                    description: "Finish and save your round"
                ),
                IntentSuggestion(
                    intentType: .statsLookup,
                    label: "View stats",
                    description: "Check your performance"
                )
            ]
        } else {
            return [
                IntentSuggestion(
                    intentType: .roundStart,
                    label: "Start round",
                    description: "Begin tracking your round"
                ),
                IntentSuggestion(
                    intentType: .equipmentInfo,
                    label: "View my bag",
                    description: "See your club setup"
                ),
                IntentSuggestion(
                    intentType: .statsLookup,
                    label: "View stats",
                    description: "Check your performance"
                )
            ]
        }
    }
}
